{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e871a7b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from glob import glob\n",
    "import os\n",
    "import pathlib\n",
    "from pathlib import Path\n",
    "import re\n",
    "import yaml\n",
    "import enum\n",
    "\n",
    "from typing import List"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "521252e1",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "incomplete input (2748383011.py, line 19)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[30], line 19\u001b[0;36m\u001b[0m\n\u001b[0;31m    def _init_tcl(self, contents: str):\u001b[0m\n\u001b[0m                                       ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m incomplete input\n"
     ]
    }
   ],
   "source": [
    "class ModuleFormat(enum.Enum):\n",
    "    TCL = enum.auto()\n",
    "    LUA = enum.auto()\n",
    "    UNKNOWN = enum.auto()\n",
    "\n",
    "class ModuleMeta:\n",
    "    \n",
    "    def __init__(self, path: Path):\n",
    "        self.path = path\n",
    "        self.module_format = ModuleFormat.UNKNOWN\n",
    "        contents = self.path.read_text()\n",
    "        if contents.startswith('#%Module1.0'):\n",
    "            self.module_format = ModuleFormat.TCL\n",
    "            self._init_tcl(contents)\n",
    "        elif contents.startswith('-- -*- lua -*-'):\n",
    "            self.module_format = ModuleFormat.LUA\n",
    "            self._init_lua(contents)\n",
    "    \n",
    "    def _init_tcl(self, contents: str):\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "fa379f05",
   "metadata": {},
   "outputs": [],
   "source": [
    "modulefiles_path = '../../spack-ucdavis/modulefiles/hpccf/franklin/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "9fdde9ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_modulefiles(roots: List[Path]):\n",
    "    for root in roots:\n",
    "        for wroot, dirs, files in os.walk(root):\n",
    "            for fn in (fn for fn in files if not fn.endswith('.yaml')):\n",
    "                module_name = os.path.join(wroot.removeprefix(root), fn)\n",
    "                print(module_name, os.path.join(wroot, fn))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "3992f1be",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'#%Module1.0\\n## Module file created by spack (https://github.com/spack/spack) on 2023-05-30 16:51:48.132371\\n##\\n## gcc@4.9.4%gcc@9.5.0~binutils+bootstrap~graphite~nvptx~piclibs~profiled~strip build_system=autotools build_type=RelWithDebInfo languages=c,c++,fortran patches=27e9829,7112abb,f5643d9,f73de87 arch=linux-ubuntu22.04-x86_64_v2/d255x4v\\n##\\n## Configure options: --with-pkgversion=Spack GCC --with-bugurl=https://github.com/spack/spack/issues --disable-multilib --enable-languages=c,c++,fortran --disable-nls --disable-canonical-system-headers --disable-libsanitizer --enable-bootstrap --with-mpfr-include=/share/apps/22.04/spack/spack-v0.19.1/opt/spack/linux-ubuntu22.04-x86_64_v2/gcc-9.5.0/mpfr-3.1.6-q4aamyikmvo4axrkxjhfjmpmw5sxkmao/include --with-mpfr-lib=/share/apps/22.04/spack/spack-v0.19.1/opt/spack/linux-ubuntu22.04-x86_64_v2/gcc-9.5.0/mpfr-3.1.6-q4aamyikmvo4axrkxjhfjmpmw5sxkmao/lib --with-gmp-include=/share/apps/22.04/spack/spack-v0.19.1/opt/spack/linux-ubuntu22.04-x86_64_v2/gcc-9.5.0/gmp-6.2.1-t7mom3odhpwv3uc3nrbkiffeplgt6bwt/include --with-gmp-lib=/share/apps/22.04/spack/spack-v0.19.1/opt/spack/linux-ubuntu22.04-x86_64_v2/gcc-9.5.0/gmp-6.2.1-t7mom3odhpwv3uc3nrbkiffeplgt6bwt/lib --with-mpc-include=/share/apps/22.04/spack/spack-v0.19.1/opt/spack/linux-ubuntu22.04-x86_64_v2/gcc-9.5.0/mpc-1.1.0-wd7gvlpl33th2ddr6iealekjqjoo6f4w/include --with-mpc-lib=/share/apps/22.04/spack/spack-v0.19.1/opt/spack/linux-ubuntu22.04-x86_64_v2/gcc-9.5.0/mpc-1.1.0-wd7gvlpl33th2ddr6iealekjqjoo6f4w/lib --without-isl --with-stage1-ldflags=-Wl,-rpath,/share/apps/22.04/spack/spack-v0.19.1/opt/spack/linux-ubuntu22.04-x86_64_v2/gcc-9.5.0/gcc-4.9.4-d255x4vbusu6tovl72zfui73i4trwpnq/lib -Wl,-rpath,/share/apps/22.04/spack/spack-v0.19.1/opt/spack/linux-ubuntu22.04-x86_64_v2/gcc-9.5.0/gcc-4.9.4-d255x4vbusu6tovl72zfui73i4trwpnq/lib64 -Wl,-rpath,/share/apps/22.04/spack/spack-v0.19.1/opt/spack/linux-ubuntu22.04-x86_64_v2/gcc-9.5.0/gmp-6.2.1-t7mom3odhpwv3uc3nrbkiffeplgt6bwt/lib -Wl,-rpath,/share/apps/22.04/spack/spack-v0.19.1/opt/spack/linux-ubuntu22.04-x86_64_v2/gcc-9.5.0/mpc-1.1.0-wd7gvlpl33th2ddr6iealekjqjoo6f4w/lib -Wl,-rpath,/share/apps/22.04/spack/spack-v0.19.1/opt/spack/linux-ubuntu22.04-x86_64_v2/gcc-9.5.0/mpfr-3.1.6-q4aamyikmvo4axrkxjhfjmpmw5sxkmao/lib --with-boot-ldflags=-Wl,-rpath,/share/apps/22.04/spack/spack-v0.19.1/opt/spack/linux-ubuntu22.04-x86_64_v2/gcc-9.5.0/gcc-4.9.4-d255x4vbusu6tovl72zfui73i4trwpnq/lib -Wl,-rpath,/share/apps/22.04/spack/spack-v0.19.1/opt/spack/linux-ubuntu22.04-x86_64_v2/gcc-9.5.0/gcc-4.9.4-d255x4vbusu6tovl72zfui73i4trwpnq/lib64 -Wl,-rpath,/share/apps/22.04/spack/spack-v0.19.1/opt/spack/linux-ubuntu22.04-x86_64_v2/gcc-9.5.0/gmp-6.2.1-t7mom3odhpwv3uc3nrbkiffeplgt6bwt/lib -Wl,-rpath,/share/apps/22.04/spack/spack-v0.19.1/opt/spack/linux-ubuntu22.04-x86_64_v2/gcc-9.5.0/mpc-1.1.0-wd7gvlpl33th2ddr6iealekjqjoo6f4w/lib -Wl,-rpath,/share/apps/22.04/spack/spack-v0.19.1/opt/spack/linux-ubuntu22.04-x86_64_v2/gcc-9.5.0/mpfr-3.1.6-q4aamyikmvo4axrkxjhfjmpmw5sxkmao/lib -static-libstdc++ -static-libgcc --with-build-config=spack\\n##\\n\\n\\nmodule-whatis \"The GNU Compiler Collection includes front ends for C, C++, Objective-C, Fortran, Ada, and Go, as well as libraries for these languages.\"\\n\\nproc ModulesHelp { } {\\nputs stderr \"The GNU Compiler Collection includes front ends for C, C++, Objective-C,\"\\nputs stderr \"Fortran, Ada, and Go, as well as libraries for these languages.\"\\n}\\n\\nconflict gcc\\n\\nprepend-path LD_LIBRARY_PATH \"/share/apps/22.04/spack/spack-v0.19.1/opt/spack/linux-ubuntu22.04-x86_64_v2/gcc-9.5.0/gcc-4.9.4-d255x4vbusu6tovl72zfui73i4trwpnq/lib64\"\\nprepend-path LD_LIBRARY_PATH \"/share/apps/22.04/spack/spack-v0.19.1/opt/spack/linux-ubuntu22.04-x86_64_v2/gcc-9.5.0/gcc-4.9.4-d255x4vbusu6tovl72zfui73i4trwpnq/lib\"\\nprepend-path PATH \"/share/apps/22.04/spack/spack-v0.19.1/opt/spack/linux-ubuntu22.04-x86_64_v2/gcc-9.5.0/gcc-4.9.4-d255x4vbusu6tovl72zfui73i4trwpnq/bin\"\\nprepend-path MANPATH \"/share/apps/22.04/spack/spack-v0.19.1/opt/spack/linux-ubuntu22.04-x86_64_v2/gcc-9.5.0/gcc-4.9.4-d255x4vbusu6tovl72zfui73i4trwpnq/share/man\"\\nprepend-path CMAKE_PREFIX_PATH \"/share/apps/22.04/spack/spack-v0.19.1/opt/spack/linux-ubuntu22.04-x86_64_v2/gcc-9.5.0/gcc-4.9.4-d255x4vbusu6tovl72zfui73i4trwpnq/.\"\\nsetenv CC \"/share/apps/22.04/spack/spack-v0.19.1/opt/spack/linux-ubuntu22.04-x86_64_v2/gcc-9.5.0/gcc-4.9.4-d255x4vbusu6tovl72zfui73i4trwpnq/bin/gcc\"\\nsetenv CXX \"/share/apps/22.04/spack/spack-v0.19.1/opt/spack/linux-ubuntu22.04-x86_64_v2/gcc-9.5.0/gcc-4.9.4-d255x4vbusu6tovl72zfui73i4trwpnq/bin/g++\"\\nsetenv FC \"/share/apps/22.04/spack/spack-v0.19.1/opt/spack/linux-ubuntu22.04-x86_64_v2/gcc-9.5.0/gcc-4.9.4-d255x4vbusu6tovl72zfui73i4trwpnq/bin/gfortran\"\\nsetenv F77 \"/share/apps/22.04/spack/spack-v0.19.1/opt/spack/linux-ubuntu22.04-x86_64_v2/gcc-9.5.0/gcc-4.9.4-d255x4vbusu6tovl72zfui73i4trwpnq/bin/gfortran\"\\nsetenv GCC_ROOT \"/share/apps/22.04/spack/spack-v0.19.1/opt/spack/linux-ubuntu22.04-x86_64_v2/gcc-9.5.0/gcc-4.9.4-d255x4vbusu6tovl72zfui73i4trwpnq\"\\n\\nif [module-info mode load] {\\n}\\n'"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Path('../../spack-ucdavis/modulefiles/hpccf/farm/spack/core/gcc/4.9.4').read_text()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "b52f8dc1",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "intel-oneapi-compilers/2022.2.1 ../../spack-ucdavis/modulefiles/hpccf/farm/spack/core/intel-oneapi-compilers/2022.2.1\n",
      "jdk/17.0.1 ../../spack-ucdavis/modulefiles/hpccf/farm/spack/core/jdk/17.0.1\n",
      "openmpi/4.1.5 ../../spack-ucdavis/modulefiles/hpccf/farm/spack/core/openmpi/4.1.5\n",
      "gcc/4.9.4 ../../spack-ucdavis/modulefiles/hpccf/farm/spack/core/gcc/4.9.4\n",
      "gcc/7.5.0 ../../spack-ucdavis/modulefiles/hpccf/farm/spack/core/gcc/7.5.0\n",
      "gcc/5.5.0 ../../spack-ucdavis/modulefiles/hpccf/farm/spack/core/gcc/5.5.0\n",
      "pmix/4.1.1 ../../spack-ucdavis/modulefiles/hpccf/farm/spack/core/pmix/4.1.1\n",
      "ucx/1.13.1 ../../spack-ucdavis/modulefiles/hpccf/farm/spack/core/ucx/1.13.1\n",
      "nvhpc/23.1 ../../spack-ucdavis/modulefiles/hpccf/farm/spack/core/nvhpc/23.1\n",
      "julia/1.8.5 ../../spack-ucdavis/modulefiles/hpccf/farm/spack/core/julia/1.8.5\n",
      "julia/1.9.0 ../../spack-ucdavis/modulefiles/hpccf/farm/spack/core/julia/1.9.0\n",
      "hwloc/2.9.0 ../../spack-ucdavis/modulefiles/hpccf/farm/spack/core/hwloc/2.9.0\n",
      "pigz/2.7 ../../spack-ucdavis/modulefiles/hpccf/farm/spack/core/pigz/2.7\n",
      "cuda/11.7.1 ../../spack-ucdavis/modulefiles/hpccf/farm/spack/core/cuda/11.7.1\n",
      "openjdk/11.0.17_8 ../../spack-ucdavis/modulefiles/hpccf/farm/spack/core/openjdk/11.0.17_8\n",
      "openjdk/16.0.2 ../../spack-ucdavis/modulefiles/hpccf/farm/spack/core/openjdk/16.0.2\n",
      "slurm/22.05.8 ../../spack-ucdavis/modulefiles/hpccf/farm/spack/core/slurm/22.05.8\n",
      "libevent/2.1.12 ../../spack-ucdavis/modulefiles/hpccf/farm/spack/core/libevent/2.1.12\n"
     ]
    }
   ],
   "source": [
    "find_modulefiles(['../../spack-ucdavis/modulefiles/hpccf/farm/spack/core/'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b6bae3c1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "StdEnv.lua\n",
      "mash/2.3.lua\n",
      "meme/5.3.0.lua\n",
      "intel-oneapi-compilers/2022.2.1.lua\n",
      "glimmer/3.02b.lua\n",
      "kraken2/2.1.2.lua\n",
      "ncbi-vdb/3.0.0.lua\n",
      "krakenuniq/0.7.3.lua\n",
      "cap3/2015-02-11.lua\n",
      "rstudio-server/2022.12.0-353.lua\n",
      "clustalw/2.1.lua\n",
      "kallisto/0.48.0.lua\n",
      "fftw/3.3.10.lua\n",
      "orthomcl/2.0.9.lua\n",
      "fastqc/0.11.9.lua\n",
      "smartdenovo/master.lua\n",
      "miniasm/2018-3-30.lua\n",
      "ea-utils/2021-10-20.lua\n",
      "sortmerna/2017-07-13.lua\n",
      "minimap2/2.14.lua\n",
      "picard/2.26.2.lua\n",
      "muscle/3.8.1551.lua\n",
      "mirdeep2/0.0.8.lua\n",
      "homer/4.9.1.lua\n",
      "nextflow/22.10.1.lua\n",
      "abyss/2.3.1.lua\n",
      "mcl/14-137.lua\n",
      "exonerate-gff3/2.3.0.lua\n",
      "parallel/20220522.lua\n",
      "mummer4/4.0.0rc1.lua\n",
      "intel-oneapi-mkl/2022.2.1.lua\n",
      "ghostscript/9.56.1.lua\n",
      "amdfftw/3.2+amd.lua\n",
      "amdfftw/default\n",
      "corset/1.09.lua\n",
      "relion-helper/0.1.lua\n",
      "relion-helper/0.2.lua\n",
      "openmpi/4.1.4.lua\n",
      "gcc/7.5.0.lua\n",
      "gcc/5.5.0.lua\n",
      "gcc/4.9.4.lua\n",
      "freebayes/1.3.6.lua\n",
      "recon/1.05.lua\n",
      "pmix/4.1.2.lua\n",
      "patchelf/0.16.1.lua\n",
      "patchelf/default\n",
      "interproscan/5.56-89.0.lua\n",
      "igv/2.12.3.lua\n",
      "ray/2.3.1.lua\n",
      "blast-plus/2.12.0.lua\n",
      "kmergenie/1.7044.lua\n",
      "ctffind/4.1.14+amd.lua\n",
      "ctffind/4.1.14+intel.lua\n",
      "ctffind/default\n",
      "scallop/0.10.5.lua\n",
      "mmseqs2/14-7e284.lua\n",
      "seqtk/1.3.lua\n",
      "repeatscout/1.0.5.lua\n",
      "hmmer/3.3.2.lua\n",
      "kalign/3.3.1.lua\n",
      "gctf/1.06.lua\n",
      "blast2go/5.2.5.lua\n",
      "ncbi-toolkit/26_0_1.lua\n",
      "bwa/0.7.17.lua\n",
      "ucx/1.13.1.lua\n",
      "repeatmasker/4.0.9.lua\n",
      "exonerate/2.4.0.lua\n",
      "infernal/1.1.4.lua\n",
      "ncbi-rmblastn/2.11.0.lua\n",
      "pilon/1.22.lua\n",
      "plink/1.07.lua\n",
      "genrich/0.6.lua\n",
      "ant/1.10.7.lua\n",
      "motioncor2/1.5.0.lua\n",
      "motioncor2/default\n",
      "orthofinder/2.5.4.lua\n",
      "phylip/3.697.lua\n",
      "hwloc/2.8.0.lua\n",
      "repeatmodeler/1.0.11.lua\n",
      "masurca/4.0.9.lua\n",
      "prokka/1.14.6.lua\n",
      "iq-tree/2.1.3.lua\n",
      "sabre/2013-09-27.lua\n",
      "cufflinks/2.2.1.lua\n",
      "hisat2/2.2.0.lua\n",
      "last/1282.lua\n",
      "sickle/1.33.lua\n",
      "bowtie/1.3.0.lua\n",
      "jellyfish/1.1.11.lua\n",
      "rnaquast/2.2.0.lua\n",
      "bwtool/1.0.lua\n",
      "minced/0.3.2.lua\n",
      "bedtools2/2.30.0.lua\n",
      "canu/2.2.lua\n",
      "cuda/11.7.1.lua\n",
      "cuda/default\n",
      "openldap/2.4.49.lua\n",
      "mummer/3.23.lua\n",
      "R/4.2.0.lua\n",
      "R/4.1.1.lua\n",
      "satsuma2/2021-03-04.lua\n",
      "relion/cpu/3.1.3+amd.lua\n",
      "relion/cpu/4.0.0+amd.lua\n",
      "relion/cpu/4.0.1+amd.lua\n",
      "relion/cpu/default\n",
      "relion/gpu/3.1.3+intel.lua\n",
      "relion/gpu/4.0.1+intel.lua\n",
      "relion/gpu/4.0.0+intel.lua\n",
      "relion/gpu/3.1.3+amd.lua\n",
      "relion/gpu/4.0.0+amd.lua\n",
      "relion/gpu/4.0.1+amd.lua\n",
      "relion/gpu/default\n",
      "clustal-omega/1.2.4.lua\n",
      "blat/35.lua\n",
      "raxml-ng/1.0.2.lua\n",
      "bowtie2/2.4.2.lua\n",
      "iqtree2/2.1.2.lua\n",
      "slurm/22-05-6-1.lua\n",
      "gatk/3.8.1.lua\n",
      "gatk/4.2.6.1.lua\n",
      "metaeuk/6-a5d39d9.lua\n",
      "mothur/1.48.0.lua\n",
      "libevent/2.1.12.lua\n",
      "r/4.2.0.lua\n",
      "kraken/1.0.lua\n",
      "seqprep/1.3.2.lua\n",
      "aragorn/1.2.38.lua\n",
      "emboss/6.6.0.lua\n"
     ]
    }
   ],
   "source": [
    "find_modulefiles(['../../spack-ucdavis/modulefiles/hpccf/franklin/Core/'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "beebe6fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_spack_index(modulefiles_path):\n",
    "    with open(os.path.join(modulefiles_path, 'module-index.yaml')) as fp:\n",
    "        module_index = yaml.load(fp, yaml.SafeLoader)['module_index']\n",
    "    return module_index\n",
    "\n",
    "def build_module_data(module_index):\n",
    "    modules = {}\n",
    "    for pkg_hash in module_index:\n",
    "        name = module_index[pkg_hash]['use_name'].split('/')\n",
    "        path = pathlib.PurePath(module_index[pkg_hash]['path'])\n",
    "        relpath = pathlib.PurePath(os.path.join(*path.parts[5:]))\n",
    "        path = modulefiles_path / relpath\n",
    "\n",
    "        fullname = os.path.join(*name)\n",
    "        software = name[0]\n",
    "        version, _, arch = name[-1].partition('+')\n",
    "        build = None if len(name) == 2 else name[1]\n",
    "\n",
    "        if software not in modules:\n",
    "            modules[software] = {}\n",
    "            modules[software]['variants'] = {}\n",
    "            modules[software]['versions'] = set()\n",
    "            modules[software]['builds'] = set()\n",
    "            modules[software]['arches'] = set()\n",
    "\n",
    "        entry = modules[software]\n",
    "        entry['versions'].add(version)\n",
    "        entry['variants'][fullname] = path\n",
    "        if build:\n",
    "            entry['builds'].add(build)\n",
    "        if arch:\n",
    "            entry['arches'].add(arch)\n",
    "    \n",
    "    return modules\n",
    "\n",
    "def render_spack_modules(modules):\n",
    "    for software, entry in sorted(modules.items()):\n",
    "        path = list(entry['variants'].values())[-1]\n",
    "        versions = ', '.join(entry['versions'])\n",
    "        variants = ', '.join(entry['builds'])\n",
    "\n",
    "        arches = entry['arches']\n",
    "        if not arches:\n",
    "            arches.add('generic')\n",
    "        arches = ', '.join(arches)\n",
    "\n",
    "        with open(path) as fp:\n",
    "            contents = fp.read()\n",
    "            about = re.search(r'help\\(\\[\\[([^\\]]*)\\]\\]\\)', contents, flags=re.MULTILINE).group(1)\n",
    "        \n",
    "        rendered = ( f'## {software}\\n\\n'\n",
    "                   f'{about}\\n\\n'\n",
    "                   f'**Versions**: {versions}\\n'\n",
    "                   f'{\"**Variants**: \" if variants else \"\"}{variants}\\n'\n",
    "                   f'**Arches**: {arches}\\n')\n",
    "        print(rendered)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "18eae41e",
   "metadata": {},
   "outputs": [],
   "source": [
    "modules = build_module_data(parse_spack_index(modulefiles_path))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "0f0fb345",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "## abyss\n",
      "\n",
      "ABySS is a de novo, parallel, paired-end sequence assembler that is\n",
      "designed for short reads. The single-processor version is useful for\n",
      "assembling genomes up to 100 Mbases in size.\n",
      "\n",
      "**Versions**: 2.3.1\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## amdfftw\n",
      "\n",
      "FFTW (AMD Optimized version) is a comprehensive collection of fast C\n",
      "routines for computing the Discrete Fourier Transform (DFT) and various\n",
      "special cases thereof. It is an open-source implementation of the Fast\n",
      "Fourier transform algorithm. It can compute transforms of real and\n",
      "complex-values arrays of arbitrary size and dimension. AMD Optimized\n",
      "FFTW is the optimized FFTW implementation targeted for AMD CPUs. For\n",
      "single precision build, please use precision value as float. Example :\n",
      "spack install amdfftw precision=float\n",
      "\n",
      "**Versions**: 3.2\n",
      "\n",
      "**Arches**: amd\n",
      "\n",
      "## aragorn\n",
      "\n",
      "ARAGORN, a program to detect tRNA genes and tmRNA genes in nucleotide\n",
      "sequences.\n",
      "\n",
      "**Versions**: 1.2.38\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## bedtools2\n",
      "\n",
      "Collectively, the bedtools utilities are a swiss-army knife of tools for\n",
      "a wide-range of genomics analysis tasks. The most widely-used tools\n",
      "enable genome arithmetic: that is, set theory on the genome.\n",
      "\n",
      "**Versions**: 2.30.0\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## blast-plus\n",
      "\n",
      "Basic Local Alignment Search Tool.\n",
      "\n",
      "**Versions**: 2.12.0\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## blast2go\n",
      "\n",
      "Blast2GO is a bioinformatics platform for high-quality functional\n",
      "annotation and analysis of genomic datasets.\n",
      "\n",
      "**Versions**: 5.2.5\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## blat\n",
      "\n",
      "BLAT (BLAST-like alignment tool) is a pairwise sequence alignment\n",
      "algorithm.\n",
      "\n",
      "**Versions**: 35\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## bowtie\n",
      "\n",
      "Bowtie is an ultrafast, memory-efficient short read aligner for short\n",
      "DNA sequences (reads) from next-gen sequencers.\n",
      "\n",
      "**Versions**: 1.3.0\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## bowtie2\n",
      "\n",
      "Bowtie 2 is an ultrafast and memory-efficient tool for aligning\n",
      "sequencing reads to long reference sequences\n",
      "\n",
      "**Versions**: 2.4.2\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## bwa\n",
      "\n",
      "Burrow-Wheeler Aligner for pairwise alignment between DNA sequences.\n",
      "\n",
      "**Versions**: 0.7.17\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## bwtool\n",
      "\n",
      "bwtool is a command-line utility for bigWig files.\n",
      "\n",
      "**Versions**: 1.0\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## canu\n",
      "\n",
      "A single molecule sequence assembler for genomes large and small.\n",
      "\n",
      "**Versions**: 2.2\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## cap3\n",
      "\n",
      "CAP3 is DNA Sequence Assembly Program\n",
      "\n",
      "**Versions**: 2015-02-11\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## clustal-omega\n",
      "\n",
      "Clustal Omega: the last alignment program you'll ever need.\n",
      "\n",
      "**Versions**: 1.2.4\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## clustalw\n",
      "\n",
      "Multiple alignment of nucleic acid and protein sequences.\n",
      "\n",
      "**Versions**: 2.1\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## corset\n",
      "\n",
      "Corset is a command-line software program to go from a de novo\n",
      "transcriptome assembly to gene-level counts.\n",
      "\n",
      "**Versions**: 1.09\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## ctffind\n",
      "\n",
      "Fast and accurate defocus estimation from electron micrographs.\n",
      "\n",
      "**Versions**: 4.1.14\n",
      "\n",
      "**Arches**: amd, intel\n",
      "\n",
      "## cuda\n",
      "\n",
      "CUDA is a parallel computing platform and programming model invented by\n",
      "NVIDIA. It enables dramatic increases in computing performance by\n",
      "harnessing the power of the graphics processing unit (GPU). Note: This\n",
      "package does not currently install the drivers necessary to run CUDA.\n",
      "These will need to be installed manually. See:\n",
      "https://docs.nvidia.com/cuda/ for details.\n",
      "\n",
      "**Versions**: 11.8.0\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## cufflinks\n",
      "\n",
      "Cufflinks assembles transcripts, estimates their abundances, and tests\n",
      "for differential expression and regulation in RNA-Seq samples.\n",
      "\n",
      "**Versions**: 2.2.1\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## ea-utils\n",
      "\n",
      "Command-line tools for processing biological sequencing data. Barcode\n",
      "demultiplexing, adapter trimming, etc. Primarily written to support an\n",
      "Illumina based pipeline - but should work with any FASTQs.\n",
      "\n",
      "**Versions**: 2021-10-20\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## emboss\n",
      "\n",
      "EMBOSS is a free Open Source software analysis package specially\n",
      "developed for the needs of the molecular biology (e.g. EMBnet) user\n",
      "community\n",
      "\n",
      "**Versions**: 6.6.0\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## exonerate\n",
      "\n",
      "Pairwise sequence alignment of DNA and proteins\n",
      "\n",
      "**Versions**: 2.4.0\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## exonerate-gff3\n",
      "\n",
      "This is an exonerate fork with added gff3 support. Original website with\n",
      "user guides: http://www.ebi.ac.uk/~guy/exonerate/\n",
      "\n",
      "**Versions**: 2.3.0\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## fastqc\n",
      "\n",
      "A quality control tool for high throughput sequence data.\n",
      "\n",
      "**Versions**: 0.11.9\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## fftw\n",
      "\n",
      "FFTW is a C subroutine library for computing the discrete Fourier\n",
      "transform (DFT) in one or more dimensions, of arbitrary input size, and\n",
      "of both real and complex data (as well as of even/odd data, i.e. the\n",
      "discrete cosine/sine transforms or DCT/DST). We believe that FFTW, which\n",
      "is free software, should become the FFT library of choice for most\n",
      "applications.\n",
      "\n",
      "**Versions**: 3.3.10\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## freebayes\n",
      "\n",
      "Bayesian haplotype-based genetic polymorphism discovery and genotyping.\n",
      "\n",
      "**Versions**: 1.3.6\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## gatk\n",
      "\n",
      " Genome Analysis Toolkit Variant Discovery in High-Throughput Sequencing\n",
      "Data\n",
      "\n",
      "**Versions**: 4.2.6.1, 3.8.1\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## gcc\n",
      "\n",
      "The GNU Compiler Collection includes front ends for C, C++, Objective-C,\n",
      "Fortran, Ada, and Go, as well as libraries for these languages.\n",
      "\n",
      "**Versions**: 4.9.4, 5.5.0, 7.5.0\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## gctf\n",
      "\n",
      "a GPU accelerated program for Real-Time CTF determination, refinement,\n",
      "evaluation and correction.\n",
      "\n",
      "**Versions**: 1.06\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## genrich\n",
      "\n",
      "Genrich is a peak-caller for genomic enrichment assays.\n",
      "\n",
      "**Versions**: 0.6\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## ghostscript\n",
      "\n",
      "An interpreter for the PostScript language and for PDF.\n",
      "\n",
      "**Versions**: 9.56.1\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## glimmer\n",
      "\n",
      "Glimmer is a system for finding genes in microbial DNA, especially the\n",
      "genomes of bacteria, archaea, and viruses.\n",
      "\n",
      "**Versions**: 3.02b\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## hisat2\n",
      "\n",
      "HISAT2 is a fast and sensitive alignment program for mapping next-\n",
      "generation sequencing reads (whole-genome, transcriptome, and exome\n",
      "sequencing data) against the general human population (as well as\n",
      "against a single reference genome).\n",
      "\n",
      "**Versions**: 2.2.0\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## hmmer\n",
      "\n",
      "HMMER is used for searching sequence databases for sequence homologs,\n",
      "and for making sequence alignments. It implements methods using\n",
      "probabilistic models called profile hidden Markov models (profile HMMs).\n",
      "\n",
      "**Versions**: 3.3.2\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## homer\n",
      "\n",
      "Software for motif discovery and next generation sequencing analysis\n",
      "\n",
      "**Versions**: 4.9.1\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## hwloc\n",
      "\n",
      "The Hardware Locality (hwloc) software project. The Portable Hardware\n",
      "Locality (hwloc) software package provides a portable abstraction\n",
      "(across OS, versions, architectures, ...) of the hierarchical topology\n",
      "of modern architectures, including NUMA memory nodes, sockets, shared\n",
      "caches, cores and simultaneous multithreading. It also gathers various\n",
      "system attributes such as cache and memory information as well as the\n",
      "locality of I/O devices such as network interfaces, InfiniBand HCAs or\n",
      "GPUs. It primarily aims at helping applications with gathering\n",
      "information about modern computing hardware so as to exploit it\n",
      "accordingly and efficiently.\n",
      "\n",
      "**Versions**: 2.8.0\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## igv\n",
      "\n",
      "The Integrative Genomics Viewer (IGV) is a high-performance\n",
      "visualization tool for interactive exploration of large, integrated\n",
      "genomic datasets. It supports a wide variety of data types, including\n",
      "array-based and next-generation sequence data, and genomic annotations.\n",
      "\n",
      "**Versions**: 2.12.3\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## infernal\n",
      "\n",
      "Infernal (INFERence of RNA ALignment) is for searching DNA sequence\n",
      "databases for RNA structure and sequence similarities. It is an\n",
      "implementation of a special case of profile stochastic context-free\n",
      "grammars called covariance models (CMs).\n",
      "\n",
      "**Versions**: 1.1.4\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## intel-oneapi-mkl\n",
      "\n",
      "Intel oneAPI Math Kernel Library (Intel oneMKL; formerly Intel Math\n",
      "Kernel Library or Intel MKL), is a library of optimized math routines\n",
      "for science, engineering, and financial applications. Core math\n",
      "functions include BLAS, LAPACK, ScaLAPACK, sparse solvers, fast Fourier\n",
      "transforms, and vector math. LICENSE INFORMATION: By downloading and\n",
      "using this software, you agree to the terms and conditions of the\n",
      "software license agreements at https://intel.ly/393CijO.\n",
      "\n",
      "**Versions**: 2022.2.1\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## interproscan\n",
      "\n",
      "InterProScan is the software package that allows sequences (protein and\n",
      "nucleic) to be scanned against InterPro's signatures. Signatures are\n",
      "predictive models, provided by several different databases, that make up\n",
      "the InterPro consortium.\n",
      "\n",
      "**Versions**: 5.56-89.0\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## iq-tree\n",
      "\n",
      "IQ-TREE Efficient software for phylogenomic inference\n",
      "\n",
      "**Versions**: 2.1.3\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## iqtree2\n",
      "\n",
      "Efficient and versatile phylogenomic software by maximum likelihood\n",
      "\n",
      "**Versions**: 2.1.2\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## jellyfish\n",
      "\n",
      "JELLYFISH is a tool for fast, memory-efficient counting of k-mers in\n",
      "DNA.\n",
      "\n",
      "**Versions**: 1.1.11\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## kalign\n",
      "\n",
      "A fast multiple sequence alignment program for biological sequences.\n",
      "\n",
      "**Versions**: 3.3.1\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## kallisto\n",
      "\n",
      "kallisto is a program for quantifying abundances of transcripts from\n",
      "RNA-Seq data.\n",
      "\n",
      "**Versions**: 0.48.0\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## kmergenie\n",
      "\n",
      "KmerGenie estimates the best k-mer length for genome de novo assembly.\n",
      "\n",
      "**Versions**: 1.7044\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## kraken\n",
      "\n",
      "Kraken is a system for assigning taxonomic labels to short DNA\n",
      "sequences, usually obtained through metagenomic studies.\n",
      "\n",
      "**Versions**: 1.0\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## kraken2\n",
      "\n",
      "Kraken2 is a system for assigning taxonomic labels to short DNA\n",
      "sequences, usually obtained through metagenomic studies.\n",
      "\n",
      "**Versions**: 2.1.2\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## krakenuniq\n",
      "\n",
      "Metagenomics classifier with unique k-mer counting for more specific\n",
      "results.\n",
      "\n",
      "**Versions**: 0.7.3\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## last\n",
      "\n",
      "LAST finds similar regions between sequences, and aligns them. It is\n",
      "designed for comparing large datasets to each other (e.g. vertebrate\n",
      "genomes and/or large numbers of DNA reads).\n",
      "\n",
      "**Versions**: 1282\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## libevent\n",
      "\n",
      "The libevent API provides a mechanism to execute a callback function\n",
      "when a specific event occurs on a file descriptor or after a timeout has\n",
      "been reached. Furthermore, libevent also support callbacks due to\n",
      "signals or regular timeouts.\n",
      "\n",
      "**Versions**: 2.1.12\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## mash\n",
      "\n",
      " Fast genome and metagenome distance estimation using MinHash.\n",
      "\n",
      "**Versions**: 2.3\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## masurca\n",
      "\n",
      "MaSuRCA is whole genome assembly software. It combines the efficiency of\n",
      "the de Bruijn graph and Overlap-Layout-Consensus (OLC) approaches.\n",
      "\n",
      "**Versions**: 4.0.9\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## mcl\n",
      "\n",
      "The MCL algorithm is short for the Markov Cluster Algorithm, a fast and\n",
      "scalable unsupervised cluster algorithm for graphs (also known as\n",
      "networks) based on simulation of (stochastic) flow in graphs.\n",
      "\n",
      "**Versions**: 14-137\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## megahit\n",
      "\n",
      "MEGAHIT: An ultra-fast single-node solution for large and complex\n",
      "metagenomics assembly via succinct de Bruijn graph\n",
      "\n",
      "**Versions**: 1.1.4\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## meme\n",
      "\n",
      "The MEME Suite allows the biologist to discover novel motifs in\n",
      "collections of unaligned nucleotide or protein sequences, and to perform\n",
      "a wide variety of other motif-based analyses.\n",
      "\n",
      "**Versions**: 5.3.0\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## metaeuk\n",
      "\n",
      "MetaEuk is a modular toolkit designed for large-scale gene discovery and\n",
      "annotation in eukaryotic metagenomic contigs.\n",
      "\n",
      "**Versions**: 6-a5d39d9\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## minced\n",
      "\n",
      "MinCED is a program to find Clustered Regularly Interspaced Short\n",
      "Palindromic Repeats (CRISPRs) in full genomes or environmental datasets\n",
      "such as metagenomes, in which sequence size can be anywhere from 100 to\n",
      "800 bp.\n",
      "\n",
      "**Versions**: 0.3.2\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## miniasm\n",
      "\n",
      "Miniasm is a very fast OLC-based de novo assembler for noisy long reads.\n",
      "\n",
      "**Versions**: 2018-3-30\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## minimap2\n",
      "\n",
      "Minimap2 is a versatile sequence alignment program that aligns DNA or\n",
      "mRNA sequences against a large reference database. Mappy provides a\n",
      "convenient interface to minimap2.\n",
      "\n",
      "**Versions**: 2.14\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## mirdeep2\n",
      "\n",
      "miRDeep2 is a completely overhauled tool which discovers microRNA genes\n",
      "by analyzing sequenced RNAs.\n",
      "\n",
      "**Versions**: 0.0.8\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## mmseqs2\n",
      "\n",
      "MMseqs2 (Many-against-Many sequence searching) is a software suite to\n",
      "search and cluster huge protein and nucleotide sequence sets\n",
      "\n",
      "**Versions**: 14-7e284\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## mothur\n",
      "\n",
      "This project seeks to develop a single piece of open-source, expandable\n",
      "software to fill the bioinformatics needs of the microbial ecology\n",
      "community.\n",
      "\n",
      "**Versions**: 1.48.0\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## motioncor2\n",
      "\n",
      "MotionCor2 is a multi-GPU program that corrects beam-induced sample\n",
      "motion recorded on dose fractionated movie stacks. It implements a\n",
      "robust iterative alignment algorithm that delivers precise measurement\n",
      "and correction of both global and non-uniform local motions at single\n",
      "pixel level, suitable for both single-particle and tomographic images.\n",
      "MotionCor2 is sufficiently fast to keep up with automated data\n",
      "collection.\n",
      "\n",
      "**Versions**: 1.5.0\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## mummer\n",
      "\n",
      "MUMmer is a system for rapidly aligning entire genomes.\n",
      "\n",
      "**Versions**: 3.23\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## mummer4\n",
      "\n",
      "MUMmer is a versatil alignment tool for DNA and protein sequences.\n",
      "\n",
      "**Versions**: 4.0.0rc1\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## muscle\n",
      "\n",
      "MUSCLE is one of the best-performing multiple alignment programs\n",
      "according to published benchmark tests, with accuracy and speed that are\n",
      "consistently better than CLUSTALW.\n",
      "\n",
      "**Versions**: 3.8.1551\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## ncbi-rmblastn\n",
      "\n",
      "RMBlast search engine for NCBI\n",
      "\n",
      "**Versions**: 2.11.0\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## ncbi-toolkit\n",
      "\n",
      "NCBI C++ Toolkit\n",
      "\n",
      "**Versions**: 26_0_1\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## ncbi-vdb\n",
      "\n",
      "The SRA Toolkit and SDK from NCBI is a collection of tools and libraries\n",
      "for using data in the INSDC Sequence Read Archives. This package\n",
      "contains the interface to the VDB.\n",
      "\n",
      "**Versions**: 3.0.0\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## nextflow\n",
      "\n",
      "Data-driven computational pipelines.\n",
      "\n",
      "**Versions**: 22.10.1\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## openmpi\n",
      "\n",
      "An open source Message Passing Interface implementation. The Open MPI\n",
      "Project is an open source Message Passing Interface implementation that\n",
      "is developed and maintained by a consortium of academic, research, and\n",
      "industry partners. Open MPI is therefore able to combine the expertise,\n",
      "technologies, and resources from all across the High Performance\n",
      "Computing community in order to build the best MPI library available.\n",
      "Open MPI offers advantages for system and software vendors, application\n",
      "developers and computer science researchers.\n",
      "\n",
      "**Versions**: 4.1.4\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## orthofinder\n",
      "\n",
      "OrthoFinder is a fast, accurate and comprehensive analysis tool for\n",
      "comparative genomics. It finds orthologues and orthogroups infers rooted\n",
      "gene trees for all orthogroups and infers a rooted species tree for the\n",
      "species being analysed. OrthoFinder also provides comprehensive\n",
      "statistics for comparative genomic analyses. OrthoFinder is simple to\n",
      "use and all you need to run it is a set of protein sequence files (one\n",
      "per species) in FASTA format.\n",
      "\n",
      "**Versions**: 2.5.4\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## orthomcl\n",
      "\n",
      "OrthoMCL is a genome-scale algorithm for grouping orthologous protein\n",
      "sequences.\n",
      "\n",
      "**Versions**: 2.0.9\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## parallel\n",
      "\n",
      "GNU parallel is a shell tool for executing jobs in parallel using one or\n",
      "more computers. A job can be a single command or a small script that has\n",
      "to be run for each of the lines in the input.\n",
      "\n",
      "**Versions**: 20220522\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## patchelf\n",
      "\n",
      "PatchELF is a small utility to modify the dynamic linker and RPATH of\n",
      "ELF executables.\n",
      "\n",
      "**Versions**: 0.16.1\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## phylip\n",
      "\n",
      "PHYLIP (the PHYLogeny Inference Package) is a package of programs for\n",
      "inferring phylogenies (evolutionary trees).\n",
      "\n",
      "**Versions**: 3.697\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## picard\n",
      "\n",
      "Picard is a set of command line tools for manipulating high-throughput\n",
      "sequencing (HTS) data and formats such as SAM/BAM/CRAM and VCF.\n",
      "\n",
      "**Versions**: 2.26.2\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## pilon\n",
      "\n",
      "Pilon is an automated genome assembly improvement and variant detection\n",
      "tool.\n",
      "\n",
      "**Versions**: 1.22\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## plink\n",
      "\n",
      "PLINK is a free, open-source whole genome association analysis toolset,\n",
      "designed to perform a range of basic, large-scale analyses in a\n",
      "computationally efficient manner.\n",
      "\n",
      "**Versions**: 1.07\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## pmix\n",
      "\n",
      "The Process Management Interface (PMI) has been used for quite some time\n",
      "as a means of exchanging wireup information needed for interprocess\n",
      "communication. However, meeting the significant orchestration challenges\n",
      "presented by exascale systems requires that the process-to-system\n",
      "interface evolve to permit a tighter integration between the different\n",
      "components of the parallel application and existing and future SMS\n",
      "solutions. PMI Exascale (PMIx) addresses these needs by providing an\n",
      "extended version of the PMI definitions specifically designed to support\n",
      "exascale and beyond environments by: (a) adding flexibility to the\n",
      "functionality expressed in the existing APIs, (b) augmenting the\n",
      "interfaces with new APIs that provide extended capabilities, (c) forging\n",
      "a collaboration between subsystem providers including resource manager,\n",
      "fabric, file system, and programming library developers, (d)\n",
      "establishing a standards-like body for maintaining the definitions, and\n",
      "(e) providing a reference implementation of the PMIx standard that\n",
      "demonstrates the desired level of scalability while maintaining strict\n",
      "separation between it and the standard itself.\n",
      "\n",
      "**Versions**: 4.1.2\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## prokka\n",
      "\n",
      "Prokka is a software tool to annotate bacterial, archaeal and viral\n",
      "genomes quickly and produce standards-compliant output files.\n",
      "\n",
      "**Versions**: 1.14.6\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## raxml-ng\n",
      "\n",
      "RAxML-NG is a phylogenetic tree inference tool which uses maximum-\n",
      "likelihood (ML) optimality criterion. Its search heuristic is based on\n",
      "iteratively performing a series of Subtree Pruning and Regrafting (SPR)\n",
      "moves, which allows to quickly navigate to the best-known ML tree.\n",
      "RAxML-NG is a successor of RAxML (Stamatakis 2014) and leverages the\n",
      "highly optimized likelihood computation implemented in libpll (Flouri et\n",
      "al. 2014).\n",
      "\n",
      "**Versions**: 1.0.2\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## ray\n",
      "\n",
      "Parallel genome assemblies for parallel DNA sequencing\n",
      "\n",
      "**Versions**: 2.3.1\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## recon\n",
      "\n",
      "RECON: a package for automated de novo identification of repeat families\n",
      "from genomic sequences.\n",
      "\n",
      "**Versions**: 1.05\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## relion\n",
      "\n",
      "RELION (for REgularised LIkelihood OptimisatioN, pronounce rely-on) is a\n",
      "stand-alone computer program that employs an empirical Bayesian approach\n",
      "to refinement of (multiple) 3D reconstructions or 2D class averages in\n",
      "electron cryo-microscopy (cryo-EM).\n",
      "\n",
      "**Versions**: 3.1.3, 4.0.0\n",
      "**Variants**: gpu, cpu\n",
      "**Arches**: amd, intel\n",
      "\n",
      "## relion-helper\n",
      "\n",
      "Utilities for Relion Cryo-EM data processing on clusters.\n",
      "\n",
      "**Versions**: 0.1, 0.2\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## repeatmasker\n",
      "\n",
      "RepeatMasker is a program that screens DNA sequences for interspersed\n",
      "repeats and low complexity DNA sequences.\n",
      "\n",
      "**Versions**: 4.0.9\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## repeatmodeler\n",
      "\n",
      "RepeatModeler is a de-novo repeat family identification and modeling\n",
      "package.\n",
      "\n",
      "**Versions**: 1.0.11\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## repeatscout\n",
      "\n",
      "RepeatScout - De Novo Repeat Finder, Price A.L., Jones N.C. and Pevzner\n",
      "P.A.\n",
      "\n",
      "**Versions**: 1.0.5\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## rnaquast\n",
      "\n",
      "Quality assessment of de novo transcriptome assemblies from RNA-Seq data\n",
      "rnaQUAST is a tool for evaluating RNA-Seq assemblies using reference\n",
      "genome and gene database. In addition, rnaQUAST is also capable of\n",
      "estimating gene database coverage by raw reads and de novo quality\n",
      "assessment using third-party software.\n",
      "\n",
      "**Versions**: 2.2.0\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## sabre\n",
      "\n",
      "Sabre is a tool that will demultiplex barcoded reads into separate\n",
      "files. It will work on both single-end and paired-end data in fastq\n",
      "format. It simply compares the provided barcodes with each read and\n",
      "separates the read into its appropriate barcode file, after stripping\n",
      "the barcode from the read (and also stripping the quality values of the\n",
      "barcode bases). If a read does not have a recognized barcode, then it is\n",
      "put into the unknown file.\n",
      "\n",
      "**Versions**: 2013-09-27\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## satsuma2\n",
      "\n",
      "Satsuma2 is an optimsed version of Satsuma, a tool to reliably align\n",
      "large and complex DNA sequences providing maximum sensitivity (to find\n",
      "all there is to find), specificity (to only find real homology) and\n",
      "speed (to accomodate the billions of base pairs in vertebrate genomes).\n",
      "\n",
      "**Versions**: 2021-03-04\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## scallop\n",
      "\n",
      "Scallop is a reference-based transcriptome assembler for RNA-seq\n",
      "\n",
      "**Versions**: 0.10.5\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## seqprep\n",
      "\n",
      "SeqPrep is a program to merge paired end Illumina reads that are\n",
      "overlapping into a single longer read.\n",
      "\n",
      "**Versions**: 1.3.2\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## seqtk\n",
      "\n",
      "Toolkit for processing sequences in FASTA/Q formats.\n",
      "\n",
      "**Versions**: 1.3\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## sickle\n",
      "\n",
      "Sickle is a tool that uses sliding windows along with quality and length\n",
      "thresholds to determine when quality is sufficiently low to trim the\n",
      "3'-end of reads and also determines when the quality is sufficiently\n",
      "high enough to trim the 5'-end of reads.\n",
      "\n",
      "**Versions**: 1.33\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## slurm\n",
      "\n",
      "Slurm is an open source, fault-tolerant, and highly scalable cluster\n",
      "management and job scheduling system for large and small Linux clusters.\n",
      "Slurm requires no kernel modifications for its operation and is\n",
      "relatively self-contained. As a cluster workload manager, Slurm has\n",
      "three key functions. First, it allocates exclusive and/or non-exclusive\n",
      "access to resources (compute nodes) to users for some duration of time\n",
      "so they can perform work. Second, it provides a framework for starting,\n",
      "executing, and monitoring work (normally a parallel job) on the set of\n",
      "allocated nodes. Finally, it arbitrates contention for resources by\n",
      "managing a queue of pending work.\n",
      "\n",
      "**Versions**: 22-05-6-1\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## smartdenovo\n",
      "\n",
      "SMARTdenovo is a de novo assembler for PacBio and Oxford Nanopore (ONT)\n",
      "data.\n",
      "\n",
      "**Versions**: master\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## sortmerna\n",
      "\n",
      "SortMeRNA is a program tool for filtering, mapping and OTU-picking NGS\n",
      "reads in metatranscriptomic and metagenomic data\n",
      "\n",
      "**Versions**: 2017-07-13\n",
      "\n",
      "**Arches**: generic\n",
      "\n",
      "## ucx\n",
      "\n",
      "a communication library implementing high-performance messaging for\n",
      "MPI/PGAS frameworks\n",
      "\n",
      "**Versions**: 1.13.1\n",
      "\n",
      "**Arches**: generic\n",
      "\n"
     ]
    }
   ],
   "source": [
    "render_spack_modules(modules)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b091f7ab",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
